homeassistant/components/generic/camera.py

```diff
@@ -170,6 +207,11 @@ def name(self):
         """Return the name of this device."""
         return self._name
 
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return str(self._still_image_url)
```

Comment:
It looks like this accepts a template?

Reply:
I originally had the Template object but it caused errors in unittest, I think because rendering the template uses async.  Converting to str there fixed it.

Reply:
It is going break for users who are relying on the template functionality without it

Reply:
This is up to the edge of my current knowledge on how templates work.  But just to make sure I understand, is the following correct?:

- Users might use a Template for the still_image_url
- The Template contains some variables that are expanded later.
- The problem is for a user who relies on those variables expanding differently.  Two different sensors might have the same template.  So this method would generate an error as a duplicate.

 I thought the unique_id has to be constant for an entity, between restarts etc.    So having a unique ID that could evaluate differently at run time would seem to be a problem.  Does that mean that variables in a unique_id are not allowed?

Querying the url to get the image would still follow the template.

Either way, if not deriving the unique_id from the still_image_url, I'm not sure what else to use (we can't access any serial numbers/mac addresses etc).

Reply:
Short of asking the user to provide a unique id, I don't have a good solution.

Let me ask for another set of eyes in case someone has a better idea.

Reply:
@bdraco To help my understanding please could you give an example of str(self._still_image_url) that would not be unique enough to use as a unique_id?

Reply:
The template can generate anything so it could be fed by the state of another entity or a url that changes every 10 minutes

Reply:
Ok, in that case, I think the unique id mustn't be based on the rendered template.  
Based on [the template class](https://github.com/python/cpython/blob/2a36b09ce7cd67503893412b53295717d66fee19/Lib/string.py#L57) having no __str__ function, 
`str(<template_obj>)` 
is probably actually calling
`repr(<template_obj>)`
Which doesn't seem good enough (I can't easily try it the moment).

[The template documentation](https://docs.python.org/3/library/string.html#string.Template.template) suggests that `<template_obj>.template` might be a better bet here, as this should be the raw template string before substitutions.

Reply:
I have just read in the [developer docs](https://developers.home-assistant.io/docs/entity_registry_index#unacceptable-sources-for-a-unique-id) that URLs are not allowed as unique identifiers.  

I'm not sure why this constraint is there but it is, and the PR currently violates this.

In some cases 'generic' can be used to just pull a static image from a web URL.  I think the URL is the only thing that reliably distinguishes 2 'generic' entities.

Reply:
Why not just use the config flow id ?

Reply:
> Why not just use the config flow id ?
Is that allowed?  I'm not sure where config_flow_id is generated and whether it is guaranteed never to be reused.

Removing all things relating to the data means it becomes hard to trigger duplicates with abort_if_already_configured().

But I'm well out of my depth understanding the underlying code here :).

Reply:
https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-of-last-resort

Reply:
@dgomes Do you mean like this?:
```
    async def _test_connection(self, info) -> (bool, str):
        """Verify that the camera data is valid before we add it."""
        if self.unique_id in self._async_current_ids():
            return False, "still_image_already_configured"
        await self.async_set_unique_id(self.unique_id)
```


Not sure why, but while testing that way, sometimes this is None, and then the tests fail due to duplicates.

I don't fully understand what is wrong with using CONF_STILL_IMAGE_URL before the jinja template is applied.

Reply:
https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-of-last-resort

You would need to pass the config entry id from (from `async_setup_entry`) into the `GenericCamera` object

Reply:
Thanks for all your help here @dgomes, @uvjustin.  Sorry, but after 2h I am completely lost now :(

@dgomes Which file do you mean?  do you mean
init.py::async_setup_entry
or
camera.py::async_setup_entry

If the latter, how do I get the config entry id?

is the `Config Entry ID` referred to in https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-of-last-resort the same thing as `config.unique_id`? if so this seems to be `None` during a configuration.yaml import, causing my tests to fail.

is it `config.flow_id`? if so, how to do I get this within camera.py::async_setup_entry?

I'm really really stuck.  I don't understand the difference between the flow_id & config_id where they come from or where they need to go!  To be honest I cant see which bits are triggered by yaml or otherwise.

Perhaps if I commit in the failing state you'll see what I mean?